lib/iris/cube.py

```diff
@@ -1632,7 +1638,38 @@ def dtype(self):
 
     @dtype.setter
     def dtype(self, dtype):
-        self._dtype = dtype
+        if dtype != self.dtype:
+            if dtype is not None:
+                if not self.has_lazy_data():
+                    emsg = 'Cube does not have lazy data, cannot set dtype.'
+                    raise ValueError(emsg)
+                dtype = np.dtype(dtype)
+                if dtype.kind != 'i':
+                    emsg = ('Can only cast lazy data to integral dtype, '
+                            'got {!r}.')
```

Comment:
We need to be strict here. We're only allowing the `dtype` to be of integral type to support the masked case. If we didn't do this, then users would abuse this to perform casting of their data.

If we do want to support casting of data, for the sake of it, then we should support that properly, say with `cube.astype(...)` or whatever. Hence, that's why we nail this `dtype` change down to only being applied to lazy, integral data.